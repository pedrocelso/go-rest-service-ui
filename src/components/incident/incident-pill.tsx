import React, { Component } from 'react';
import { connect } from 'react-redux'
import { deauthenticate } from '../login/login-actions'
import { Task, TaskService } from '../../services/task';

interface IncidentPillProps {
  task: Task;
  service: TaskService;
  deauthenticate: typeof deauthenticate
}

interface IncidentPillState {
  loading: boolean
}

export class IncidentPill extends Component<IncidentPillProps, IncidentPillState> {
  state = {
    loading: false
  }
  
  getIncidents(taskId: number) {
    return () => {
      this.setState({loading: true})
      const {service} = this.props
      service.getIncidents(taskId)
        .fork(
          (e) => {
            this.setState({loading: false})
            if (e.statusCode === 401) {
              this.props.deauthenticate();
              sessionStorage.removeItem(`jwtToken`)
            }
          },
          (incidentList) => {
            this.setState({loading: false})
            console.log(incidentList)
          }
        )
    }
  }

  render() {
    const {task} = this.props
    const {loading} = this.state

    const element = loading ? (
      <div className="spinner-border" role="status">
        <span className="sr-only">Loading...</span>
      </div>) : (<span className="badge badge-primary badge-pill cursor-pointer" onClick={this.getIncidents(task.id)}>{task.incidentsCount}</span>)

    return element
  }
}

const mapDispatchToProps = { deauthenticate }
export default connect(null, mapDispatchToProps)(IncidentPill)