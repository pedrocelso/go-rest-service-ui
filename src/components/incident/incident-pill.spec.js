import React from 'react';
import {shallow} from 'enzyme'
import sinon from 'sinon'

import {IncidentPill} from './incident-pill'

const baseTask = {
  id: 123,
  name: `POI quest service`,
  description: `Do A POI quest service`,
  creationTime: 1556924666404,
  updateTime: 1556924566404,
  incidentsCount: 55
}

describe(`<IncidentPill />`, () => {
  it(`Should render the pill with the right incident count`, () => {
    const service = {
      getIncidents: sinon.fake()
    }

    const component = shallow(
      <IncidentPill task={baseTask} service={service} />,
    );

    expect(component).toMatchSnapshot();
  });

  it(`Should render the loading icon when service is loading`, () => {
    const service = {
      getIncidents: sinon.fake()
    }

    const component = shallow(
      <IncidentPill task={baseTask} service={service} />,
    );

    component.instance().setState({loading: true})

    expect(component).toMatchSnapshot();
  });
});