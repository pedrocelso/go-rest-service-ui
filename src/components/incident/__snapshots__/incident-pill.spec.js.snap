// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IncidentPill /> Should render the loading icon when service is loading 1`] = `
<div
  className="spinner-border"
  role="status"
>
  <span
    className="sr-only"
  >
    Loading...
  </span>
</div>
`;

exports[`<IncidentPill /> Should render the pill with the right incident count 1`] = `
<span
  className="badge badge-primary badge-pill cursor-pointer"
  onClick={[Function]}
>
  55
</span>
`;
